### PR TITLE
prep, model and train the interactions model

### PR DESCRIPTION
**Pull request by rfl-urbaniak**
Add the first version of `prep_data_for_interaction_inference()` (will be revised once we have invertible transformation in place).

Add `train_interactions_model()`.

Add `cities_model_interactions()`.

Why? First step towards building the guide-generating model training pipeline.

A brief test is in the experimental notebook named `interaction_cities_prototype.ipynb`. Ignore all other new stuff in the experimental notebooks folder.

### CODE REVIEW

**Review by riadas — APPROVED** [review]
Added some plotting libraries needed for `interaction_cities_prototype.ipynb` to run. Looked over the rest (`cities/modeling`) and looks good to me -- for the sake of efficiency, I'm fine with postponing any deeper reviews I do to when I integrate with the frontend, after the additional small PRs you make @rfl-urbaniak. Approving so you can merge!

File: cities/modeling/modeling_utils.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        geo_subsets.append(geo_subset)
+
+    outcome_df = pd.concat(geo_subsets).reset_index(drop=True)
+
+    outcome = outcome_df[
+        (outcome_df["Year"] >= year_min)
+        & (outcome_df["Year"] <= year_max + forward_shift)
+    ]
+
+    intervention = dg.std_long[intervention_dataset][
```
Comment (azane) [inline cities/modeling/modeling_utils.py:90]: maybe comment/explain this year-line-up stuff.

File: cities/modeling/modeling_utils.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    N_s = len(state_index.unique())  # number of states
+
+    assert len(intervention) == len(y)
+    assert len(unit_index) == len(y)
+    assert len(state_index) == len(unit_index)
+    assert len(time_index) == len(unit_index)
+    assert covariates.shape[1] == covariates_sparse.shape[1]
+    assert len(unit_index_sparse) == N_u
+
+    return {
```
Comment (azane) [inline cities/modeling/modeling_utils.py:184]: I don't love returning giant dicts but also don't have a better idea. Leave for now probably.

**Review by azane — APPROVED** [review]
lgtm

**Pull request merged**